### PR TITLE
Added ONNX support

### LINKED ISSUE #125: ONNX/Tensorflow Port
I just wanted to inform everybody that I was working on porting the PyTorch code to work with other deep-learning frameworks. I am doing this because PyTorch doesn't work well with other python packages for bundling into standalone programs. Also, PyTorch is large compared to other packages like ONNX when all we are doing is inference. The changes are minor and restricted to the `network.py` file.

We could make it an option for which framework to use if people have trouble installing PyTorch or want to run this package on a system that cannot run PyTorch.

### PR DESCRIPTION
**Pull request by jacobf18**
PR Description
--------------

Added ONNX support to iclabel. I changed the function `run_iclabel` to include a library parameter. This parameter's default value is set to use PyTorch. I added unit tests to check the network outputs against MATLAB. ONNX is faster to load and more lightweight than PyTorch.

The options for the `library` parameter are `pytorch` and `onnx` so far. It throws a `ValueError` for any other input.

Closes #125.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-connectivity/blob/main/doc/whats_new.rst) is updated
- [ ] New contributors have been added to [CITATION.cff](https://github.com/mne-tools/mne-icalabel/blob/main/CITATION.cff)
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

File: mne_icalabel/iclabel/tests/test_network.py [hunk 1]
```diff
@@ -126,14 +128,58 @@ def test_network_outputs():
...
+
 @pytest.mark.parametrize(
     "eeglab_feature_file, eeglab_feature_formatted_file",
     [
         (features_raw_path, features_formatted_raw_path),
         (features_epo_path, features_formatted_epo_path),
     ],
 )
-def test_format_input(eeglab_feature_file, eeglab_feature_formatted_file):
+def test_format_input_pytorch(eeglab_feature_file, eeglab_feature_formatted_file):
```
Comment (mscheltienne) [inline mne_icalabel/iclabel/tests/test_network.py:182]:
```suggestion
def test_format_input(eeglab_feature_file, eeglab_feature_formatted_file):
```

This is testing a function common to both. I would keep the old name.

File: mne_icalabel/iclabel/tests/test_network.py [hunk 2]
```diff
@@ -126,14 +128,58 @@ def test_network_outputs():
...
+    # - elements in position [0, 2, 4] -> 1-element array with the var name
+    # - elements in position [1, 3, 5] -> data arrays
+    assert "in_image" == features[0][0]
+    images = features[1]
+    assert "in_psdmed" == features[2][0]
+    psd = features[3]
+    assert "in_autocorr" == features[4][0]
+    autocorr = features[5]
+
+    # reshape the features to fit torch format
```
Comment (mscheltienne) [inline mne_icalabel/iclabel/tests/test_network.py:153]:
```suggestion
    # reshape the features to fit onnx format
```

File: pyproject.toml [hunk 3]
```diff
@@ -56,7 +57,7 @@ doc = [
     'memory_profiler',
     'numpydoc',
     'pydata-sphinx-theme',
-    'PyQt5',
+    'PyQt6',
```
Comment (mscheltienne) [inline pyproject.toml:60]: Any reason to change this?

**Review by mscheltienne — APPROVED** [review]
Thanks @jacobf18 ! That looks good to me.

As you said, we are going to need to add some logic to load the libraries. I'm also +1 to make Qt an optional dependency. Overall, +1 to merge as is and clean-up, add library selection logic, and so on.. in a second PR.

File: pyproject.toml [hunk 4]
```diff
@@ -42,6 +42,7 @@ dependencies = [
     'mne >= 1.1',
     'pooch',
     'torch',
+    'onnxruntime',
```
Comment (adam2392) [inline pyproject.toml:45]: Is there a way we can make both optional dependencies, since the user only needs to install one?

**Review by adam2392 — COMMENTED** [review]
If we can make the dependencies optional, then we should nest the import statements for torch and onnx inside the functions that use them. 

Perhaps we should add a helper function that does the import and checks that at least one of them is installed?

**Comment by mscheltienne** [comment]
Yes, we need to add some selection logic and get rid of the import at the top-level. Probably requires also 2 separate files, one for `pytorch` and one for `onnx` since the classes defined inherit from `torch.nn` directly. Maybe some logic like `psychopy` has [here](https://github.com/psychopy/psychopy/blob/81dc09ae935515ea43d97fc5eef281457d3a8c83/psychopy/parallel/__init__.py#L34-L57) in an `__init__`.
Anyway, I think that could be done in a second PR which restructure the `iclabel` module before a 0.5 release.

**Comment by adam2392** [comment]
Alright. Sounds good to me. LGTM with the exception of CI. Does it just need a restart possibly?

Thanks @jacobf18!

**Comment by mscheltienne** [comment]
No the ubuntu ones have been broken for a while because of some xvfb stuff, it will be a good occasion to fix them as well..

**Comment by mscheltienne** [comment]
@jacobf18 Do you want to give a shot to a logic to set `pytorch` and `onnx` as optional?
Ideally:
- `import mne_icalabel` and `from mne_icalabel import label_components` would work regardless of the installation status of the backends.
- `import mne_icalabel.iclabel` and `from mne_icalabel.iclabel import label_components` would fail if neither backend is available.

A couple of ideas: we should move the network part to a `mne_icalabel.iclabel.network` module, with 3 files: 
- `__init__.py` which attempts to load both in a similar fashion to psychopy with the drivers for the parallel port
- `pytorch.py` with the pytorch backend
- `onnx.py` with the onnx backend

Or maybe an empty `__init__.py` and some helper function which handles the loading of the backends.
The tests might also require a decorator `@requires_onnx` or `@requires_torch`.

-------

Let me know if you want to give it a shot, else I'll do it in the coming days. Once it's sorted out, we can release a 0.5 version with this addition, and maybe get this correctly packaged on `conda-forge` for windows.

**Comment by mscheltienne** [comment]
Merging and picking this up today with another PR to add the selection of the backend. Let's not stall this :)

**Pull request merged**